nmdc_schema/migration_recursion.py

```diff
@@ -283,11 +292,114 @@ def __init__(self) -> None:
 
     def force_research_study_study_category(self, study: dict) -> dict:
         if 'study_category' not in study:
-            logger.info(f"Forcing 'study_category: research_study' on {study['id']}")
+            logger.info(
+                f"Forcing 'study_category: research_study' on {study['id']}")
             study['study_category'] = 'research_study'
         return study
 
 
+class Migrator_from_8_1_0_to_9_0_0(MigratorBase):
+    """Migrates data from schema 8.1.0 to 9.0.0"""
+
+    def __init__(self) -> None:
+        """Invokes parent constructor and populates collection-to-transformations map."""
+
+        super().__init__()
+
+        # Populate the "collection-to-transformers" map for this specific migration.
+        self.agenda = dict(
+            study_set=[self.fix_award_dois, self.fix_pub_dois,
+                       self.fix_massive, self.fix_ess_dive, self.remove_doi_slots],
+        )
+
+    def process_doi(self, study: dict, doi_list, doi_category):
+        id_value = study['id']
+        for doi_updates in doi_list:
+            if id_value == doi_updates['id']:
+                new_doi = {
+                    'doi_value': doi_updates['doi'],
+                    'doi_category': doi_category,
+                    'doi_provider': doi_updates['doi_prov']
+                }
+                study.setdefault('associated_dois', []).append(new_doi)
```

Comment:
Note to self:

`study['associated_dois']` is a list of dictionaries. Here, we are appending a new dictionary to the list.